runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp

```diff
@@ -322,8 +322,174 @@ J9::ARM64::TreeEvaluator::DIVCHKEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+   TR::LabelSymbol *snippetLabel;
+   TR::Instruction *gcPoint;
+   TR::Snippet *snippet;
+
+   TR::addDependency(deps, obj1Reg, TR::RealRegister::NoReg, TR_GPR, cg);
+   TR::addDependency(deps, obj2Reg, TR::RealRegister::NoReg, TR_GPR, cg);
+   TR::addDependency(deps, tmp1Reg, TR::RealRegister::NoReg, TR_GPR, cg);
+   TR::addDependency(deps, tmp2Reg, TR::RealRegister::NoReg, TR_GPR, cg);
+
+   // Same array, we are done.
+   generateCompareInstruction(cg, node, obj1Reg, obj2Reg);
+   generateConditionalBranchInstruction(cg, TR::InstOpCode::b_cond, node, doneLabel, TR::CC_EQ);
+
+   // If we know nothing about either object, test object1 first. It has to be an array.
+   if ((!node->isArrayChkPrimitiveArray1())
+         && (!node->isArrayChkReferenceArray1())
+         && (!node->isArrayChkPrimitiveArray2())
+         && (!node->isArrayChkReferenceArray2()))
+      {
+#ifdef OMR_GC_COMPRESSED_POINTERS
+      generateTrg1MemInstruction(cg, TR::InstOpCode::ldrw, node, tmp1Reg,
+            new (cg->trHeapMemory()) TR::MemoryReference(obj1Reg, (int32_t) TR::Compiler->om.offsetOfObjectVftField(), cg));
+#else
+      generateTrg1MemInstruction(cg, isLong ? TR::InstOpCode::ldrx : TR::InstOpCode::ldrw, node, tmp1Reg,
+            new (cg->trHeapMemory()) TR::MemoryReference(obj1Reg, TR::Compiler->om.offsetOfObjectVftField(), cg));
+#endif
+
+      TR::TreeEvaluator::generateVFTMaskInstruction(cg, node, tmp1Reg);
+      generateTrg1MemInstruction(cg, TR::InstOpCode::ldrw, node, tmp1Reg,
+            new (cg->trHeapMemory()) TR::MemoryReference(tmp1Reg, offsetof(J9Class, classDepthAndFlags), cg));
+
+      loadConstant32(cg, node, J9AccClassRAMArray, tmp2Reg);
+      generateTrg1Src2Instruction(cg, TR::InstOpCode::andw, node, tmp2Reg, tmp1Reg, tmp2Reg);
+
+      generateCompareImmInstruction(cg, node, tmp2Reg, NULLVALUE);
+
+      snippetLabel = generateLabelSymbol(cg);
+      gcPoint = generateConditionalBranchInstruction(cg, TR::InstOpCode::b_cond, node, snippetLabel, TR::CC_EQ);
+      snippet = new (cg->trHeapMemory()) TR::ARM64HelperCallSnippet(cg, node, snippetLabel, node->getSymbolReference(), doneLabel);
+      cg->addSnippet(snippet);
+      }
+
+   // One of the object is array. Test equality of two objects' classes.
+#ifdef OMR_GC_COMPRESSED_POINTERS
+   // must read only 32 bits
+   generateTrg1MemInstruction(cg, TR::InstOpCode::ldrw, node, tmp2Reg,
+         new (cg->trHeapMemory()) TR::MemoryReference(obj2Reg, (int32_t) TR::Compiler->om.offsetOfObjectVftField(), cg));
+   generateTrg1MemInstruction(cg, TR::InstOpCode::ldrw, node, tmp1Reg,
+         new (cg->trHeapMemory()) TR::MemoryReference(obj1Reg, (int32_t) TR::Compiler->om.offsetOfObjectVftField(), cg));
+#else
+   generateTrg1MemInstruction(cg, isLong ? TR::InstOpCode::ldrx : TR::InstOpCode::ldrw, node, tmp2Reg,
+         new (cg->trHeapMemory()) TR::MemoryReference(obj2Reg, TR::Compiler->om.offsetOfObjectVftField(), TR::Compiler->om.sizeofReferenceAddress(), cg));
```

Comment:
There is no `MemoryReference` constructor available of type `(TR::Register *, int, int, TR::CodeGenerator *)`.